[feat] add course events to calendar

mumer92 opened the pull request:
### Description

[LEARNER-8301](https://openedx.atlassian.net/browse/LEARNER-8301)

As a user, I want to add all course events to my device's calendar, for a self-paced course

### How to test this PR

- [x]   An "add to calendar" message and CTA (switch) is shown on the Course Dates page
- [x] Audit track users see the message, To complete graded assignments as part of this course, you can upgrade today.
- [ ] Verified track users see a different message, We built a suggested schedule to help you stay on track. But don’t worry – it’s flexible so you can learn at your own pace. If you happen to fall behind, you’ll be able to adjust the dates to keep yourself on track.
- [x] Message/CTA is shown only for self-paced, and hidden for instructor-paced courses
- [ ] Remote flags exist to show/hide the message and CTA
- [x] Switching a CTA from off to on for the first time prompts the user for permission to modify their calendar
- [x] If the user does not allow permission, switching a CTA from off to on notifies the user that the app does not have permission, and directs them to Settings to change their app permissions
- [ ] Android experience varies a bit, should be the standard experience by platform
- [x] Switching the CTA from off to on from the Course Dates page presents a native alert or modal to confirm adding the course events:
- Title: Add calendar “Technical Support Fundamentals”
- Body: Would you like to add the edX calendar `<Calendar Name>`? You can edit or remove your course calendar any time in Calendar or Settings.
- [x] Confirming adding course events adds all events found on the Course Dates page to the device's calendar. Events are visible in the calendar.
- [x] Added course events
- [x] Calendar events are added to a course specific calendar named edX - <Course Name>
- [x] The calendar "edX" is created if it does not already exist (sub calendar is created on device with course name)
- [x] Are 1 hour in duration
- [x] Start 1 hour before the due time
- [x] Have the name Assignment Due: <Course Name>
- [x] Have a description with the same text as in the Dates tab, without the deep link (deep link is in a different story).  For example: “Homework 6 Peer Assessment (Submission)” or “MA1 section graded exam questions: MA1 Section 3 graded exam questions”
- [x] Switching the CTA from on to off from the Course Dates page deletes the previously added events
- [x] Analytics events

saeedbashir commented on Source/CourseDatesViewController.swift:
```diff
@@ -37,17 +37,90 @@ class CourseDatesViewController: UIViewController, InterfaceOrientationOverridin
     
+    private lazy var calendar: CalendarManager = {
+        guard let course = course, let courseName = course.name else {
+            return CalendarManager(courseName: "EDX")
```
Can we use platform_name instead of hardcoded name?

saeedbashir commented on Source/CourseDatesViewController.swift:
```diff
@@ -37,17 +37,90 @@ class CourseDatesViewController: UIViewController, InterfaceOrientationOverridin
+                return
+            }
+                        
+            UserDefaults.standard.setValue(newValue, forKey: courseName)
```
instead of adding all the names into the user defaults directly, how about grouping them under a key like calendar entries?

saeedbashir commented on Source/CourseDatesViewController.swift:
```diff
@@ -83,7 +156,97 @@ class CourseDatesViewController: UIViewController, InterfaceOrientationOverridin
+        }
+    }
+    
+    private func addEventsToCalendar() {
```
It seems this method is not adding events to the calendar so it can be renamed according to its functionality.

saeedbashir commented on Source/CourseDatesViewController.swift:
```diff
@@ -83,7 +156,97 @@ class CourseDatesViewController: UIViewController, InterfaceOrientationOverridin
+        }
+    }
+    
+    private func addCourseEvents(completion: (()->())? = nil) {
+        calendar.addEventsToCalendar(for: dateBlocksMap) { [weak self] done, error in
+            if done {
+                self?.trackCalendarEvent(for: .CalendarAddDatesSuccess, eventName: .CalendarAddDatesSuccess)
+                completion?()
+            }
+        }
+    }
+    
+    private func removeEventsFromCalendar(completion: (()->())? = nil) {
+        calendar.removeCalendar { [weak self] done, _ in
+            if done {
+                self?.trackCalendarEvent(for: .CalendarRemoveDatesSuccess, eventName: .CalendarRemoveDatesSuccess)
+                completion?()
+            }
+        }
+    }
```
How about moving calendar-related code to an extension?

saeedbashir commented on Source/CourseDatesViewController.swift:
```diff
@@ -120,6 +284,7 @@ class CourseDatesViewController: UIViewController, InterfaceOrientationOverridin
             case .success((var courseDateModel, let userPreference)):
                 if courseDateModel.dateBlocks.isEmpty {
                     self?.loadController.state = .failed(message: Strings.Coursedates.courseDateUnavailable)
+                    self?.calendar.removeCalendar()
```
You are removing the calendar on dates shift, are you also reading the events on the dates shift?

mumer92 replied:
I added this if a course has finished its running and dates api is called, if it had calendar entries, it should remove it.

mumer92 replied:
We will be handling shifting in coming ticket.

saeedbashir commented on Source/en.lproj/Localizable.strings:
```diff
@@ -316,6 +316,26 @@
+/* Course Dates Add Calendar Dialog Title */
+"COURSEDATES.ADD_CALENDAR_TITLE"="Add Calendar \"{calendar_name}\"";
+/* Course Dates Add Calendar Message prompt */
+"COURSEDATES.ADD_CALENDAR_PROMPT"="Would you like to add the edX Calendar \"{calendar_name}\" ? \n You can edit or remove the course calendar any time in Calendar or Settings";
```
instead of hardcoded `edX` can you please read it from the config file?

saeedbashir reviewed (changes requested):
* After adding the events to the calendar if a user removes the calendar access from the settings then turning off the sync isn't deleting the calendar events. We need to decide what message we should show to the user.
* Design is not exactly in party with the Figma design, font sizes and spacing are not the same.
* Figma design is using a different snack bar for success messages.

mumer92 commented on Source/CourseDatesHeaderView.swift:
```diff
@@ -0,0 +1,342 @@
+        return label
+    }()
+    
+    private lazy var buttonContainer = UIView()
```
yes, actually in stackView, the container's width is equal to stackView's width. So adding button inside container and giving it a width do works. If i were to remove container and give width to button, it does not works and takes up full width available to it.

![Simulator Screen Shot - iPhone 12 Pro - 2021-05-05 at 19 30 31](https://user-images.githubusercontent.com/960241/117158214-a94eff80-add8-11eb-9805-e526ce91c487.png)

mumer92 commented on Source/CourseDatesHeaderView.swift:
```diff
@@ -0,0 +1,342 @@
+        }
+    }
+    
+    private var estimatedTopContainerHeight: Int {
+        return estimatedHeightForTitleLabel + estimatedHeightForMessageLabel + estimatedButtonHeight
+    }
+    
+    private var estimatedBottomContainerHeight: Int {
+        return estimatedHeightForCalendarMessageLabel + estimatedHeightForSwitchContainer + (Int(StandardVerticalMargin) * 2)
+    }
+    
+    var estimatedHeight: Int {
+        return isSelfPaced ? estimatedTopContainerHeight + estimatedBottomContainerHeight : estimatedTopContainerHeight
+    }
```
I have tested it, but seems like we do need to give height to header, if height is not given then it all becomes grabbled. please check following screenshot

![Simulator Screen Shot - iPhone 12 Pro - 2021-05-05 at 19 42 53](https://user-images.githubusercontent.com/960241/117160146-4eb6a300-adda-11eb-8e4e-ab5214e5826c.png)

mumer92 commented:
> * After adding the events to the calendar if a user removes the calendar access from the settings then turning off the sync isn't deleting the calendar events. We need to decide what message we should show to the user.

I think we need to handle this in a followup ticket and show some message to user that we are unable to delete events because calendar is not found.

> * Design is not exactly in party with the Figma design, font sizes and spacing are not the same.

I have addressed design changes

> * Figma design is using a different snack bar for success messages.

I have added a new snackbar for calendar actions.

saeedbashir reviewed (changes requested):
* On denying the calendar access app is crashing.
* Successful addition of calendar should have only one button `Done` because cancel and done both are just hiding the alert.
* After adding and deleting the entries for a course on resyncing the calendar isn't working.

saeedbashir reviewed (changes requested):
* Don't allow access to the calendar opens settings alert. It should only open when the user didn't allow it previously and then later try to sync the course dates.
* Turning on the sync and then cancels the next alert results in enabling the switch. To see it on come back from the enrollment screen.

mumer92 commented on Source/CalendarManager.swift:
```diff
@@ -0,0 +1,274 @@
+    }
+    
+    private func generateCalendarEvent(for block: CourseDateBlock) -> EKEvent {
+        let title = block.title + ": " + courseName
+        let startDate = block.blockDate.addingTimeInterval(startDateOffsetHour * 60 * 60)
+        let endDate = block.blockDate
+        let notes = "\(courseName) \n \(block.title)"
```
actually i placed them in 2 separate methods as the notes are made different for a date that is single and for the dates that are on same time.

for single block date at given time 
`let notes = "\(courseName) \n \(block.title)"` 
for multiple block dates at given time 
`let notes = "\(courseName) \n" + blocks.compactMap { $0.title }.joined(separator: ", ")`

saeedbashir reviewed (approved):
LGTM 👍
